### PR TITLE
#527 Keychain admins management feature

### LINKED ISSUE #527: Add messages for updating Keychains admins
At this time, when a Keychain is created, the creator is the sole admin. There is no way to add new admins, or remove an admin.
We should add new messages to do that. To get started, let's say that all admins are equal and can add or remove any other admin, including themselves.

Note: ensure that a Keychain always has at least one admin.

### PR DESCRIPTION
**Pull request by backsapc**
- Added MsgAddKeychainAdminRequest, MsgRemoveKeychainAdminRequest commands
- Added EventAddKeychainAdmin, EventRemoveKeychainAdmin events

https://github.com/warden-protocol/wardenprotocol/issues/527

## Summary by CodeRabbit

- **New Features**
	- Introduced the ability to add and remove keychain administrators, enhancing administrative capabilities.
	- New event types for adding and removing keychain admins to improve event management.
	- Added methods for dynamic management of administrator roles within the keychain.
- **Bug Fixes**
	- Improved validations for admin management operations to ensure proper authority checks.
- **Documentation**
	- Updated the changelog to reflect new features and functionalities.

### CODE REVIEW

**Review by artur-abliazimov — DISMISSED** [review]

File: proto/warden/warden/v1beta3/events.proto [hunk 1]
```diff
@@ -170,3 +170,29 @@ message EventAddKeychainWriter {
   // updated count of writers
   uint64 writers_count = 3;
 }
+
+// EventAddKeychainWriter is emitted when a new writer address is added to a Keychain
+message EventAddKeychainAdmin {
+  // id of the keychain
+  uint64 id = 1;
+
+  // address of the new writer
```
Comment (pitasi) [inline proto/warden/warden/v1beta3/events.proto]:
```suggestion
  // address of the new admin
```

File: proto/warden/warden/v1beta3/events.proto [hunk 2]
```diff
@@ -170,3 +170,29 @@ message EventAddKeychainWriter {
...
+  string new_admin = 2;
+
+  // updated count of writers
```
Comment (pitasi) [inline proto/warden/warden/v1beta3/events.proto]:
```suggestion
  // updated count of admins
```

File: proto/warden/warden/v1beta3/events.proto [hunk 3]
```diff
@@ -170,3 +170,29 @@ message EventAddKeychainWriter {
   // updated count of writers
   uint64 writers_count = 3;
 }
+
+// EventAddKeychainWriter is emitted when a new writer address is added to a Keychain
```
Comment (pitasi) [inline proto/warden/warden/v1beta3/events.proto]:
```suggestion
// EventAddKeychainAdmin is emitted when a new admin address is added to a Keychain
```

File: proto/warden/warden/v1beta3/events.proto [hunk 4]
```diff
@@ -170,3 +170,29 @@ message EventAddKeychainWriter {
...
+
+  // address of the new writer
+  string new_admin = 2;
+
+  // updated count of writers
+  uint64 admins_count = 3;
+}
+
+
+// EventAddKeychainWriter is emitted when a new writer address is added to a Keychain
```
Comment (pitasi) [inline proto/warden/warden/v1beta3/events.proto]:
```suggestion
// EventAddKeychainAdmin is emitted when a new admin address is added to a Keychain
```

File: proto/warden/warden/v1beta3/events.proto [hunk 5]
```diff
@@ -170,3 +170,29 @@ message EventAddKeychainWriter {
...
+message EventRemoveKeychainAdmin {
+  // id of the keychain
+  uint64 id = 1;
+
+  // address of the new writer
```
Comment (pitasi) [inline proto/warden/warden/v1beta3/events.proto]:
```suggestion
  // address of the new admin
```

File: proto/warden/warden/v1beta3/events.proto [hunk 6]
```diff
@@ -170,3 +170,29 @@ message EventAddKeychainWriter {
...
+  string admin = 2;
+
+  // updated count of writers
```
Comment (pitasi) [inline proto/warden/warden/v1beta3/events.proto]:
```suggestion
  // updated count of admins
```

File: proto/warden/warden/v1beta3/tx.proto [hunk 7]
```diff
@@ -153,6 +153,24 @@ message MsgUpdateKeychain {
 
 message MsgUpdateKeychainResponse {}
 
+message MsgAddKeychainAdminRequest {
+  option (cosmos.msg.v1.signer) = "authority";
+  string authority = 1;
+  uint64 keychain_id = 2;
+  string new_admin = 3;
```
Comment (pitasi) [inline proto/warden/warden/v1beta3/tx.proto]:
```suggestion
  string new_admin = 3 [(cosmos_proto.scalar) = "cosmos.AddressString"];
```

we should start using this annotation for strings that are addresses, at some point we'll have to fix all the other places too 😅

File: warden/x/warden/keeper/msg_server_add_keychain_admin.go [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+	kr, err := k.keychains.Get(ctx, msg.KeychainId)
+	if err != nil {
+		return nil, err
+	}
+
+	if !kr.IsAdmin(msg.Authority) {
+		return nil, fmt.Errorf("keychain updates should be requested by admins")
+	}
+	if kr.IsAdmin(msg.NewAdmin) {
+		return &types.MsgAddKeychainAdminResponse{}, nil
```
Comment (pitasi) [inline warden/x/warden/keeper/msg_server_add_keychain_admin.go]: let's return an error here so the user knows that the request was wrong

File: warden/x/warden/keeper/msg_server_remove_keychain_admin.go [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+	kr, err := k.keychains.Get(ctx, msg.KeychainId)
+	if err != nil {
+		return nil, err
+	}
+
+	if !kr.IsAdmin(msg.Authority) {
+		return nil, fmt.Errorf("keychain updates should be requested by admins")
+	}
+	if !kr.IsAdmin(msg.Admin) {
+		return &types.MsgRemoveKeychainAdminResponse{}, nil
```
Comment (pitasi) [inline warden/x/warden/keeper/msg_server_remove_keychain_admin.go]: as above, let's return an error here

**Review by pitasi — CHANGES_REQUESTED** [review]
i left some comments, nothing big, basically just return errors if the context is not what we expect and a few copy-paste mistakes 😁

**Comment by backsapc** [comment]
@Pitasi fixed all of the comments

**Review by pitasi — APPROVED** [review]

**Pull request merged**